src/node_file.cc

```diff
@@ -84,19 +84,26 @@ const char* const kPathSeparator = "\\/";
 #endif
 
 std::string Basename(const std::string& str, const std::string& extension) {
-  std::string ret = str;
-
   // Remove everything leading up to and including the final path separator.
-  std::string::size_type pos = ret.find_last_of(kPathSeparator);
-  if (pos != std::string::npos) ret = ret.substr(pos + 1);
+  std::string::size_type pos = str.find_last_of(kPathSeparator);
+
+  // Starting index for the resulting string
+  std::size_t start_pos = 0;
+  // String size to return
+  std::size_t str_size = str.size();
+  if (pos != std::string::npos) {
+    start_pos = pos + 1;
+    str_size -= start_pos;
+  }
 
   // Strip away the extension, if any.
-  if (ret.size() >= extension.size() &&
-      ret.substr(ret.size() - extension.size()) == extension) {
-    ret = ret.substr(0, ret.size() - extension.size());
+  if (str_size >= extension.size() &&
+      // Because we still have the whole string intact
+      str.substr(start_pos + str_size - extension.size()) == extension) {
```

Comment:
If the goal is to remove `str.substr` (presumably because it allocated memory), why not replace this with `str.compare`? Or does the compiler optimize this `substr` call away?

Reply:
Not sure about the implementation of compare and how it handles the comparison and allocation but from a higher level point of view it seems valid.

Reply:
`compare` returns an integer. If the strings are equal, the return value is `0`.